src/test/UniswapV2LockstakeCallee.t.sol

```diff
@@ -0,0 +1,418 @@
+
+interface MkrAuthorityLike {
+    function rely(address) external;
+}
+
+contract MockUniswapRouter02 is DssTest {
+    uint256 fixedPrice;
+
+    constructor(uint256 price_) {
+        fixedPrice = price_;
+    }
+
+    // Hardcoded to simulate fixed price Uniswap
+    /* uniRouter02.swapExactTokensForTokens(gemAmt, daiToJoin + minProfit, path, address(this), block.timestamp); */
+    function swapExactTokensForTokens(uint amountIn, uint amountOutMin, address[] calldata path, address to, uint deadline) external returns (uint[] memory amounts) {
+        to; deadline; // silence warning
+        uint buyAmt = amountIn * fixedPrice;
+        require(amountOutMin <= buyAmt, "Minimum Fill not reached");
+
+        uint256 initialInBalance = DSTokenAbstract(path[0]).balanceOf(address(this));
+        DSTokenAbstract(path[0]).transferFrom(msg.sender, address(this), amountIn);
+        assertEq(DSTokenAbstract(path[0]).balanceOf(address(this)), initialInBalance + amountIn);
+
+        uint256 initialOutBalance = DSTokenAbstract(path[path.length - 1]).balanceOf(msg.sender);
+        GodMode.setBalance(path[path.length - 1], msg.sender, initialOutBalance + buyAmt);
+        assertEq(DSTokenAbstract(path[path.length - 1]).balanceOf(msg.sender), initialOutBalance + buyAmt);
+
+        amounts = new uint[](2);
+        amounts[0] = amountIn;
+        amounts[1] = buyAmt;
+    }
+
+}
+
+contract UniswapV2LockstakeCalleeTest is DssTest {
+    using stdStorage for StdStorage;
+
+    DssInstance             dss;
+    address                 pauseProxy;
+    DSTokenAbstract         dai;
+    DSTokenAbstract         mkr;
+    LockstakeMkr            lsmkr;
+    LockstakeEngine         engine;
+    LockstakeClipper        clip;
+    address                 calc;
+    MedianAbstract          pip;
+    VoteDelegateFactoryMock voteDelegateFactory;
+    NstMock                 nst;
+    NstJoinMock             nstJoin;
+    GemMock                 rTok;
+    StakingRewardsMock      farm;
+    StakingRewardsMock      farm2;
+    MkrNgtMock              mkrNgt;
+    GemMock                 ngt;
+    bytes32                 ilk = "LSE";
+    address                 voter;
+    address                 voteDelegate;
+
+    LockstakeConfig     cfg;
+
+    uint256             prevLine;
+    
+    address constant LOG = 0xdA0Ab1e0017DEbCd72Be8599041a2aa3bA7e740F;
+
+    MockUniswapRouter02      uniRouter02;
+    uint256                  uniV2Price = 1;
+    address[]                uniV2Path = new address[](2);
+    UniswapV2LockstakeCallee callee;
+
+    event OnKick(address indexed urn, uint256 wad);
+    event OnTake(address indexed urn, address indexed who, uint256 wad);
+    event OnRemove(address indexed urn, uint256 sold, uint256 burn, uint256 refund);
+
+    // Match https://github.com/makerdao/lockstake/blob/735e1e85ca706534a77d8e1582df0d3248cbd2b6/test/LockstakeEngine.t.sol#L87-L177
```

Comment:
nit, 
it doesn't fully match with the commented link as `dai` is missing there. 

https://www.diffchecker.com/5eGEigBq/